# Influence of the nature and wellbeing of a Country in its Co2 emissions

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

### Data sets
This analysis is based on date extracted from [Gapminder](https://www.google.com/url?q=http://www.gapminder.org/data/&sa=D&ust=1532469042121000), a website that has collected a lot of
information about how people live their lives in dierent countries, tracked across the years, and on a number of diferent indicators. 

I've focused my analysis on enviromental pollution and what influences it.

The analysis consists of the following indicators:
 
- [C02 Emission (tonnes per person)](https://cdiac.ess-dive.lbl.gov/)
Carbon dioxide emissions from the burning of fossil fuels (metric tonnes of C02 per person). 
This indicator has been chosen to analyze the ecological behaviour of each country, the more Co2 emissions a country makes the less ecological friendly it is.  

- [Forest coverage(%)](https://www.fao.org/forestry/sofo/en/)
Percentage of total land area that has been covered with forest during the given year.
This indicator has been chosen to analyze the amount of nature that surrounds the people in each country, and also to see its evolution accross the years. 

- [Democracy score](http://www.systemicpeace.org/inscrdata.html)
Summary measure of a country's democratic and free nature in all independent countries with total population greater than 500,000 in 2018. 
For a better understanding of this index, -10 is the lowest value and 10 the highest. 
This indicator has been chosen to understand the influece of the people on the countries' decisions. 
_This dataset was extracted directly from the sorucelink, since it was a newer version than the one provided in gapminder._

- [Human Development Index (HDI)](http://hdr.undp.org/en/indicators/137506)
Index used to rank countries by level of "human development". It contains three dimensions: health level, educational level and living standard. 


### Objective

The main purpose of this analysis is to understand if the wellbeing of the society (HDI), the democratic system (polity) or the surrounding nature (forest coverage) impact somehow its ecological impact (emissions of Co2). 
Do countries with a better democratic system, understood as a more influence of the people in its decisions, less Co2 emissions? Within these countries, does the wellbeing have an impact? Do people with a better education, health system and living standard impact on its countries ecological impact?
And lastly, has the amount of nature that surrounds these people impact on their actions? 

I'm interested in analyze the (most) current situation, and the relation between the indicators in different countries.

With the last variable, forest coverage, I also want to see if there is any relation between the increase of Co2 emissions and the decrease of the forest coverage. 

In [1]:
# Import packages
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

<a id='wrangling'></a>
## Data Wrangling


### General Properties

#### C02 Emission (tonnes per person) 

In [2]:
# Import the csv file as a dataframe
df_co2 = pd.read_csv('Data/co2_emissions_tonnes_per_person.csv')

#print first 10 lines
df_co2.head(10)

country  1800  1801  1802  1803  1804  1805  1806    1807  \
0          Afghanistan   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
1              Albania   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
2              Algeria   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
3              Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
4               Angola   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
5  Antigua and Barbuda   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
6            Argentina   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
7              Armenia   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
8            Australia   NaN   NaN   NaN   NaN   NaN   NaN   NaN     NaN   
9              Austria   NaN   NaN   NaN   NaN   NaN   NaN   NaN  0.0517   

   1808   ...       2005     2006     2007    2008    2009    2010    2011  \
0   NaN   ...     0.0529   0.0637   0.0854   0.154   0.242   0.294   0.412   
1   NaN   ...     1.3800   1.2800   1.3000   1.460   1.480   1.560   1.790   
2   NaN   ...     3.2200   2.9900   3.1900   3.160   3.420   3.300   3.290   
3   NaN   ...     7.3000   6.7500   6.5200   6.430   6.120   6.120   5.870   
4   NaN   ...     0.9800   1.1000   1.2000   1.180   1.230   1.240   1.250   
5   NaN   ...     4.8100   4.9100   5.1400   5.190   5.450   5.540   5.360   
6   NaN   ...     4.1400   4.4300   4.3800   4.680   4.410   4.560   4.600   
7   NaN   ...     1.4600   1.4800   1.7300   1.910   1.510   1.470   1.710   
8   NaN   ...    17.3000  17.8000  17.8000  18.100  18.200  17.700  17.400   
9   NaN   ...     8.9900   8.7100   8.3900   8.280   7.490   8.030   7.690   

    2012    2013    2014  
0   0.35   0.316   0.299  
1   1.68   1.730   1.960  
2   3.46   3.510   3.720  
3   5.92   5.900   5.830  
4   1.33   1.250   1.290  
5   5.42   5.360   5.380  
6   4.57   4.460   4.750  
7   1.98   1.900   1.900  
8  17.00  16.100  15.400  
9   7.31   7.280   6.800  

[10 rows x 216 columns]

In [3]:
# print shape to see the amount of countries (x) and amount of years (y)
df_co2.shape

(192, 216)

In [4]:
# Count missing values
# I'll check only the last 20 years of the dataset, because I want to use the most recent data
df_co2[df_co2.columns[-20:]].isna().sum()

1995    3
1996    3
1997    3
1998    3
1999    3
2000    3
2001    3
2002    2
2003    2
2004    2
2005    2
2006    2
2007    1
2008    1
2009    1
2010    1
2011    1
2012    0
2013    0
2014    0
dtype: int64

In [5]:
# Check the missing values
df_co2[df_co2[df_co2.columns[-20:]].isna().any(axis=1)]

country  1800  1801  1802  1803  1804  1805  1806  1807  1808  \
96   Liechtenstein   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
156    South Sudan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
171    Timor-Leste   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

     ...     2005   2006   2007   2008   2009   2010   2011   2012   2013  \
96   ...      NaN    NaN  1.760  1.750  1.440  1.530  1.310  1.300  1.390   
156  ...      NaN    NaN    NaN    NaN    NaN    NaN    NaN  0.123  0.130   
171  ...    0.171  0.171  0.172  0.187  0.208  0.212  0.217  0.254  0.372   

      2014  
96   1.190  
156  0.130  
171  0.387  

[3 rows x 216 columns]

In [6]:
# Check for duplicated rows
df_co2.duplicated().sum()

0

Even though the dataset consists of a very wide range of years, I need only the most recent data. I'll compare the latest year of each dataset to define the year that I'll use for the analysis. 
But from this exploration, the data won't need much cleaning because it has almost no missing values and no duplicates. 



#### Human Development Index (HDI)

In [7]:
# Import the csv file as a dataframe
df_hdi = pd.read_csv('Data/hdi_human_development_index.csv')

#print first 10 lines
df_hdi.head(10)


country   1990   1991   1992   1993   1994   1995   1996  \
0          Afghanistan  0.295  0.300  0.309  0.305  0.300  0.324  0.328   
1              Albania  0.635  0.618  0.603  0.608  0.616  0.628  0.637   
2              Algeria  0.577  0.581  0.587  0.591  0.595  0.600  0.609   
3              Andorra    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4               Angola    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
5  Antigua and Barbuda    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
6            Argentina  0.705  0.713  0.720  0.725  0.728  0.731  0.738   
7              Armenia  0.634  0.628  0.595  0.593  0.597  0.603  0.609   
8            Australia  0.866  0.867  0.871  0.874  0.876  0.885  0.888   
9              Austria  0.794  0.798  0.804  0.806  0.812  0.816  0.819   

    1997   1998  ...     2006   2007   2008   2009   2010   2011   2012  \
0  0.332  0.335  ...    0.415  0.433  0.434  0.448  0.454  0.463  0.470   
1  0.636  0.646  ...    0.703  0.713  0.721  0.725  0.738  0.752  0.759   
2  0.617  0.627  ...    0.690  0.697  0.705  0.714  0.724  0.732  0.737   
3    NaN    NaN  ...      NaN    NaN    NaN    NaN  0.819  0.819  0.843   
4    NaN    NaN  ...    0.454  0.468  0.480  0.488  0.495  0.508  0.523   
5    NaN    NaN  ...    0.781  0.786  0.788  0.783  0.782  0.778  0.781   
6  0.746  0.753  ...    0.788  0.792  0.794  0.802  0.816  0.822  0.823   
7  0.618  0.632  ...    0.707  0.721  0.725  0.720  0.729  0.732  0.736   
8  0.891  0.894  ...    0.918  0.921  0.925  0.927  0.927  0.930  0.933   
9  0.823  0.833  ...    0.860  0.864  0.870  0.872  0.880  0.884  0.887   

    2013   2014   2015  
0  0.476  0.479  0.479  
1  0.761  0.762  0.764  
2  0.741  0.743  0.745  
3  0.850  0.857  0.858  
4  0.527  0.531  0.533  
5  0.782  0.784  0.786  
6  0.825  0.826  0.827  
7  0.739  0.741  0.743  
8  0.936  0.937  0.939  
9  0.892  0.892  0.893  

[10 rows x 27 columns]

In [8]:
# print shape to see the amount of countries (x) and amount of years (y)
df_hdi.shape

(187, 27)

This result is already very small compared to the Co2 emissions data set that consists of records from the last 216 years compared to 27 in this one. This will not be an issue, because only the last year is needed. There's also a difference in that value, the most recent year in this dataset is 2015 but in the Co2 emissions dataset was 2014. 
I'll check the next datasets, but so far **the year of analysis will be 2014**. 

There's also a **difference in the amount of countries** that should also be compared and unified before proceding with the analysis phase. 

In [9]:
# Count missing values in the last 21 years of the dataset
# since this dataset has one more year, and I want to see the data for the same range
df_hdi[df_hdi.columns[-21:]].isna().sum()

1995    40
1996    40
1997    40
1998    40
1999    37
2000    20
2001    20
2002    20
2003    18
2004    15
2005     6
2006     6
2007     6
2008     6
2009     6
2010     0
2011     0
2012     0
2013     0
2014     0
2015     0
dtype: int64

In [10]:
# Check for duplicated rows
df_hdi.duplicated().sum()

0

In [11]:
# Check the missing values
df_hdi[df_hdi[df_hdi.columns[-21:]].isna().any(axis=1)]

country  1990  1991  1992  1993  1994  1995  1996  \
3                           Andorra   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4                            Angola   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
5               Antigua and Barbuda   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
11                          Bahamas   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
19                           Bhutan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
21           Bosnia and Herzegovina   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
26                     Burkina Faso   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
31                       Cape Verde   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
33                             Chad   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
37                          Comoros   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
48                         Dominica   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
53                Equatorial Guinea   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
54                          Eritrea   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
56                         Ethiopia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
62                          Georgia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
66                          Grenada   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
69                    Guinea-Bissau   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
87                         Kiribati   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
92                          Lebanon   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
94                          Liberia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
96                    Liechtenstein   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
99                       Madagascar   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
108           Micronesia, Fed. Sts.   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
111                      Montenegro   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
121                         Nigeria   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
122                 North Macedonia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
124                            Oman   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
126                           Palau   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
127                       Palestine   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
144                      Seychelles   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
149                 Solomon Islands   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
152                     South Sudan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
155             St. Kitts and Nevis   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
156                       St. Lucia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
157  St. Vincent and the Grenadines   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
159                        Suriname   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
167                     Timor-Leste   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
173                    Turkmenistan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
180                      Uzbekistan   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
181                         Vanuatu   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

     1997  1998  ...     2006   2007   2008   2009   2010   2011   2012  \
3     NaN   NaN  ...      NaN    NaN    NaN    NaN  0.819  0.819  0.843   
4     NaN   NaN  ...    0.454  0.468  0.480  0.488  0.495  0.508  0.523   
5     NaN   NaN  ...    0.781  0.786  0.788  0.783  0.782  0.778  0.781   
11    NaN   NaN  ...    0.790  0.791  0.791  0.788  0.788  0.789  0.790   
19    NaN   NaN  ...      NaN    NaN    NaN    NaN  0.572  0.581  0.589   
21    NaN   NaN  ...    0.703  0.710  0.716  0.717  0.711  0.728  0.735   
26    NaN   NaN  ...    0.334  0.345  0.356  0.365  0.377  0.384  0.392   
31    NaN   NaN  ...    0.602  0.615  0.621  0.627  0.632  0.636  0.643   
33    NaN   NaN  ...    0.306  

Even though there are no duplicates, the data in the oldest years has several missing values, more than 20% of the countries in the dataset are missing data on the first years. 

For a better understanding of this index, I will label the scores using the interpretation table that can be found in the indicator documenattion. 

| Human development | Score   |
|-----------------------------|-------------|
| Very high | >= 0.800    |
| High | 0.700–0.799 |
| Medium | 0.550–0.699 |
| Low | < 0.550     |

I will check this values against the the 25th, 50th, and 75th percentile to evaluate if it represents the dataset. 

In [12]:
# I'll use he column from the year 2014 because it's the latest year that I'll use for the analysis
df_hdi[['country','2014']].describe()

2014
count  187.000000
mean     0.695706
std      0.154876
min      0.347000
25%      0.570000
50%      0.724000
75%      0.804500
max      0.948000

The values are similar, I'll use the table from the instructions to label the data.

#### Forest coverage(%)

In [13]:
# Import the csv file as a dataframe
df_fc = pd.read_csv('Data/forest_coverage_percent.csv')

#print first 10 lines
df_fc.head(10)

country   1990   1991   1992   1993   1994   1995   1996  \
0          Afghanistan   2.07   2.07   2.07   2.07   2.07   2.07   2.07   
1              Albania  28.80  28.70  28.60  28.60  28.50  28.40  28.40   
2              Algeria   0.70   0.70   0.69   0.69   0.69   0.68   0.68   
3              Andorra  34.00  34.00  34.00  34.00  34.00  34.00  34.00   
4               Angola  48.90  48.80  48.70  48.60  48.50  48.40  48.30   
5  Antigua and Barbuda  23.40  23.30  23.30  23.20  23.10  23.10  23.00   
6            Argentina  12.70  12.60  12.50  12.40  12.30  12.20  12.10   
7              Armenia    NaN    NaN  11.80  11.80  11.70  11.70  11.70   
8            Australia  16.70  16.70  16.70  16.70  16.80  16.80  16.80   
9              Austria  45.70  45.80  45.90  46.00  46.00  46.10  46.20   

    1997   1998  ...     2006   2007   2008   2009   2010   2011   2012  \
0   2.07   2.07  ...     2.07   2.07   2.07   2.07   2.07   2.07   2.07   
1  28.30  28.20  ...    28.50  28.50  28.40  28.40  28.30  28.30  28.30   
2   0.67   0.67  ...     0.68   0.71   0.74   0.77   0.81   0.81   0.81   
3  34.00  34.00  ...    34.00  34.00  34.00  34.00  34.00  34.00  34.00   
4  48.20  48.10  ...    47.30  47.20  47.10  47.00  46.90  46.80  46.70   
5  22.90  22.90  ...    22.30  22.30  22.30  22.30  22.30  22.30  22.30   
6  12.00  11.90  ...    10.90  10.80  10.70  10.60  10.40  10.30  10.20   
7  11.70  11.70  ...    11.70  11.70  11.60  11.60  11.60  11.60  11.60   
8  16.80  16.80  ...    16.50  16.40  16.30  16.10  16.00  16.10  16.10   
9  46.30  46.30  ...    46.70  46.70  46.70  46.70  46.80  46.80  46.80   

    2013   2014   2015  
0   2.07   2.07   2.07  
1  28.20  28.20  28.20  
2   0.81   0.82   0.82  
3  34.00  34.00  34.00  
4  46.60  46.50  46.40  
5  22.30  22.30  22.30  
6  10.10  10.00   9.91  
7  11.70  11.70  11.70  
8  16.20  16.20  16.20  
9  46.80  46.90  46.90  

[10 rows x 27 columns]

In [14]:
# print shape to see the amount of countries (x) and amount of years (y)
df_fc.shape

(192, 27)

In this case the amount of countries matches with the Co2 indicator, but some values can still not be matching. 
The range of years matches the HDI dataset, and presents no problem with keeping the year 2014 for the analysis. 

In [15]:
# Count missing values in the last 21 years of the dataset
df_fc[df_fc.columns[-21:]].isna().sum()

1995    4
1996    4
1997    4
1998    4
1999    4
2000    2
2001    2
2002    2
2003    2
2004    2
2005    2
2006    0
2007    0
2008    0
2009    0
2010    0
2011    1
2012    1
2013    1
2014    1
2015    1
dtype: int64

In [16]:
# Check the missing values
df_fc[df_fc[df_fc.columns[-20:]].isna().any(axis=1)]

country  1990  1991  1992  1993  1994  1995  1996  1997  1998  ...   \
16      Belgium   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
98   Luxembourg   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
112  Montenegro   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
147      Serbia   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...    
162       Sudan  12.9  12.9  12.8  12.7  12.6  12.6  12.5  12.4  12.3  ...    

     2006  2007  2008  2009  2010  2011  2012  2013  2014  2015  
16   22.3  22.4  22.4  22.4  22.5  22.5  22.5  22.5  22.6  22.6  
98   33.5  33.5  33.5  33.5  33.5  33.5  33.5  33.5  33.5  33.5  
112  49.5  52.5  55.5  58.5  61.5  61.5  61.5  61.5  61.5  61.5  
147  28.9  29.4  29.9  30.5  31.0  31.0  31.1  31.1  31.1  31.1  
162  11.8  11.7  11.6  11.5  11.5   NaN   NaN   NaN   NaN   NaN  

[5 rows x 27 columns]

Since there's only one country with missing data for the year 2014 out of the 192 countries that contain the dataset, I'll drop this country from the dataset. With the rest of the countries, missing data in the years before 2005 I'll compare with the rest of the datasets and then refill the data. 

In [17]:
# Check for duplicated rows
df_fc.duplicated().sum()

0

#### Democracy score (Polity)

This dataset was obtained from the official website, because it was more updated that the version provided by gapminder. 
The main difference is found in the format of the file, it's provided as an .xls format instead of .csv, contains more variables and the year is listed as one more column instead of being the first row of the dataset. 

In [18]:
# Import the excel file as a dataframe
df_polity = pd.read_excel('Data/p4v2018.xls')

#print first 10 lines
df_polity.head(10)

cyear  ccode scode        country  year  flag  fragment  democ  autoc  \
0  21800      2   USA  United States  1800     0       NaN      7      3   
1  21801      2   USA  United States  1801     0       NaN      7      3   
2  21802      2   USA  United States  1802     0       NaN      7      3   
3  21803      2   USA  United States  1803     0       NaN      7      3   
4  21804      2   USA  United States  1804     0       NaN      7      3   
5  21805      2   USA  United States  1805     0       NaN      7      3   
6  21806      2   USA  United States  1806     0       NaN      7      3   
7  21807      2   USA  United States  1807     0       NaN      7      3   
8  21808      2   USA  United States  1808     0       NaN      7      3   
9  21809      2   USA  United States  1809     0       NaN      9      0   

   polity    ...     interim  bmonth  bday   byear  bprec  post  change   d4  \
0       4    ...         NaN     1.0   1.0  1800.0    1.0   4.0    88.0  1.0   
1       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
2       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
3       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
4       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
5       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
6       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
7       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
8       4    ...         NaN     NaN   NaN     NaN    NaN   NaN     NaN  NaN   
9       9    ...         NaN     3.0   5.0  1809.0    1.0   9.0     5.0  1.0   

   sf  regtrans  
0 NaN       NaN  
1 NaN       NaN  
2 NaN       NaN  
3 NaN       NaN  
4 NaN       NaN  
5 NaN       NaN  
6 NaN       NaN  
7 NaN       NaN  
8 NaN       NaN  
9 NaN       2.0  

[10 rows x 36 columns]

In [19]:
# The indicator we need, and that was listed in Gapminder is polity2
# Only the columns Country, year and polity2 are needed
df_polity = df_polity[['country','year','polity2']]
df_polity.sample(10)

country  year  polity2
13068        Iran  1893    -10.0
6786      Austria  2005     10.0
13672       Syria  2011     -7.0
17097   Indonesia  1985     -7.0
2068   Costa Rica  1996     10.0
15593       Japan  1859     -8.0
14841  Uzbekistan  1991     -9.0
5948     Portugal  2017     10.0
8213     Bulgaria  1887     -5.0
7834   Yugoslavia  1956     -7.0

In [20]:
# the dataset is formatted diferently than the others, so it will have to be adapted to the format
# check columns in the dataset
df_polity.info();

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17562 entries, 0 to 17561
Data columns (total 3 columns):
country    17562 non-null object
year       17562 non-null int64
polity2    17325 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 411.7+ KB


In [21]:
# Due to the difference in format I can't use shape to determine amount of years and countries
df_polity.nunique() 

country    195
year       219
polity2     21
dtype: int64

In [22]:
# Get year range
print(df_polity['year'].min(),"-", df_polity['year'].max())

1800 - 2018


This dataset has more recent data than the others, I'll have to limit this for the analysis. 
It also has more countries than the other dataframes, this will also be reduced once the list of countries gets unified. 

In [23]:
# Check for duplicated rows
df_polity.duplicated().sum()

1

In [24]:
# check the duplicated line
df_polity[df_polity.duplicated()]

country  year  polity2
7897  Yugoslavia  1991     -5.0

Even though there's a line duplicated, the problem is that the dataset has countries, like Yugoslavia that just appeared, that don't exist anymore. And this will not be a problem after keeping only the most recent records for each indicator. 

I'll check within the year 2014, which is the most recent year that has data in all datasets, and therefore will be used for the analysis. 

In [25]:
# Check amount of countries within the year range
df_polity.query('year == 2014').count()

country    167
year       167
polity2    166
dtype: int64

The amount of countries is less than the total amount of countries in the dataset, but still more than what I expect to have after the cleaning. 

There's only one record missing. 

In [26]:
#Check missing data within the year range
df_polity.query('year > 1994 and year < 2015').isna().sum()

country     0
year        0
polity2    52
dtype: int64

In [27]:
# Check the missing values
df_polity[df_polity.isna().any(axis=1)].query('year > 1994 and year < 2015').nunique()

country     6
year       20
polity2     0
dtype: int64

It's not so many countries with data missing (6 out of the 167) in the dataset.

After assessing the data there are the steps for cleaning the data will be:

**Formatting the data**
In order to analyze the data together I need to merge those dataframes and create one single dataframe where each indicator is a column. 
The expected output from the merged dataframes should have this format:

| Country   |year | Co2   | forest_coverage | HDI | polity2 |
|-----------|-----|-------|-----------------|-----|---------|
|     ...   | ... |  ...  |    ...          | ... |    ...  |


**Unifying list of countries and years**
Every dataframe has a different range of years and some countries differ from one another. All these values should be consistent accross dataframes. To accomplish that the merge between the dataframes should be an inner join, that way only countries and years existing in all dataframes will remain. 
The expected output will be the data belonging to the year range 1995-2014, that is present in all data sets. 

**Fill missing values**
Every dataframe has several data missing, but most of it will be dismissed during the inner join of the dataframes because it belongs to older years.

For visualization purposes, I'd like to keep most of the records instead of dropping the ones with missing data in order to have the first and the last observed value during the period. 

For the values that remain missing after the merge, I'll refill them using the forward fill method. This way the values will be consistent along time. 

With the remaining missing values, dropping them would leave an incosistent dataset that could generate fake outliers in the analysis, and similar would be filling with the mean, given that all of the indicadors can present a notable evolution over time, specially with the polity2 indicator. 


### Data Cleaning

#### Formatting the data and unifying list of countries and years 

In [28]:
# unpivot the data
df_co2 = df_co2.melt(id_vars=['country'], var_name=['year'])

# Use the indicator as a column name
df_co2.rename(columns={"value": "co2"}, inplace=True)

# Check the new structure
df_co2.head(10)

country  year  co2
0          Afghanistan  1800  NaN
1              Albania  1800  NaN
2              Algeria  1800  NaN
3              Andorra  1800  NaN
4               Angola  1800  NaN
5  Antigua and Barbuda  1800  NaN
6            Argentina  1800  NaN
7              Armenia  1800  NaN
8            Australia  1800  NaN
9              Austria  1800  NaN

In [29]:
# check the new structure columns
df_co2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41280 entries, 0 to 41279
Data columns (total 3 columns):
country    41280 non-null object
year       41280 non-null object
co2        16905 non-null float64
dtypes: float64(1), object(2)
memory usage: 967.6+ KB


In [30]:
# unpivot the data
df_fc = df_fc.melt(id_vars=['country'], var_name=['year'])

# Use the indicator as a column name
df_fc.rename(columns={"value": "forest_coverage"}, inplace=True)

# Check the new structure
df_fc.head(10)

country  year  forest_coverage
0          Afghanistan  1990             2.07
1              Albania  1990            28.80
2              Algeria  1990             0.70
3              Andorra  1990            34.00
4               Angola  1990            48.90
5  Antigua and Barbuda  1990            23.40
6            Argentina  1990            12.70
7              Armenia  1990              NaN
8            Australia  1990            16.70
9              Austria  1990            45.70

In [31]:
# check the new structure columns
df_fc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4992 entries, 0 to 4991
Data columns (total 3 columns):
country            4992 non-null object
year               4992 non-null object
forest_coverage    4882 non-null float64
dtypes: float64(1), object(2)
memory usage: 117.1+ KB


In [32]:
# unpivot the data
df_hdi = df_hdi.melt(id_vars=['country'], var_name=['year'])

# Use the indicator as a column name
df_hdi.rename(columns={"value": "human_development"}, inplace=True)

# Check the new structure
df_hdi.head(10)# check the new structure columns

country  year  human_development
0          Afghanistan  1990              0.295
1              Albania  1990              0.635
2              Algeria  1990              0.577
3              Andorra  1990                NaN
4               Angola  1990                NaN
5  Antigua and Barbuda  1990                NaN
6            Argentina  1990              0.705
7              Armenia  1990              0.634
8            Australia  1990              0.866
9              Austria  1990              0.794

In [33]:
# check the new structure columns
df_hdi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4862 entries, 0 to 4861
Data columns (total 3 columns):
country              4862 non-null object
year                 4862 non-null object
human_development    4322 non-null float64
dtypes: float64(1), object(2)
memory usage: 114.0+ KB


In [34]:
# Merge the dataframes in one
# co2, hdi and forest coverage can be easily merged because they have the same datatypes
df_combined = df_co2.merge(df_fc, how ='inner', on = ['country','year'])
df_combined = df_combined.merge(df_hdi, how ='inner', on = ['country','year'])

In [35]:
# melt function converted the year to a string value
# but in the polity2 dataframe it's type is int, because it was already formatted in the source file 
# change the type of the column to int to match polity and merge
df_combined['year'] = df_combined['year'].astype('int')
df_combined = df_combined.merge(df_polity, how ='inner', on = ['country','year'])

# Check the combined dataframe
df_combined.sample(10)

country  year     co2  forest_coverage  human_development  polity2
707      Turkey  1994  2.7200            12.80              0.596      8.0
2065  Argentina  2004  4.0700            11.20              0.780      8.0
1942    Comoros  2003  0.2260            23.20                NaN      4.0
3701    Tunisia  2014  2.5900             6.63              0.723      7.0
218    Mongolia  1991  5.4800             8.02              0.573      2.0
358        Mali  1992  0.0500             5.35              0.233      7.0
3305      Egypt  2012  2.4700             0.07              0.681     -3.0
2587    Lesotho  2007  1.0500             1.43              0.447      8.0
2499     Uganda  2006  0.0859            16.50              0.442     -1.0
2422  Indonesia  2006  1.5000            53.60              0.638      8.0

In [36]:
# Keep records from the last 20 years only (1995-2004)
df_combined = df_combined.query('year>1994')

In [37]:
# count countries
df_combined.nunique() 

country               151
year                   20
co2                  1564
forest_coverage       963
human_development     653
polity2                21
dtype: int64

In [38]:
# Check size
df_combined.shape

(2995, 6)

The size is not divisible by the amount of years, which means that some countries don't have all idicators for one or more years. I'll identify them and drop them. 

In [39]:
#Group by year and count contries per year
df_combined.groupby('year').size()


year
1995    149
1996    149
1997    149
1998    149
1999    149
2000    149
2001    149
2002    149
2003    149
2004    149
2005    149
2006    151
2007    151
2008    151
2009    151
2010    151
2011    151
2012    150
2013    150
2014    150
dtype: int64

There are at least 2 countries that are missing all indicators for the whole range of years, I'll drop thoise countries. 

In [40]:
# get countries that appear less than the amount of years
exc_cuntries = df_combined.country.value_counts().loc[lambda x: x<20].index.tolist()
exc_cuntries

['Sudan', 'Serbia', 'Montenegro']

In [41]:
# Keep countries that are not in the list I just create
df_combined = df_combined[~df_combined.country.isin(exc_cuntries)]

In [42]:
# Check size, it should have (151 countries - 3 excluded)*20 years = 2960 rows
df_combined.shape

(2960, 6)

I will now check the missing data to have an idea how much is still missing. 

In [43]:
#check missing data
df_combined.isna().sum()

country                0
year                   0
co2                    0
forest_coverage       10
human_development    148
polity2               31
dtype: int64

Now I'll refill the missing values with a backward fill, this method takes the propagates the non-null values backward. To avoid mixing data within countries I'll order the dataframe first.

In [44]:
df_combined.sort_values(by=['country', 'year'])
df_combined.fillna(method='bfill', inplace =True)

In [45]:
#check that all missing data has been filled
df_combined.isna().sum()

country              0
year                 0
co2                  0
forest_coverage      0
human_development    0
polity2              0
dtype: int64

In the case of the HDI data i'll label the countries depending on their score. 
To define the numbers where to "cut" the groups I'll use the ones from the table in the indicator's intructions plus the min and max value calculated. 

In [46]:
# Bin edges that will be used to "cut" the data into groups
bin_edges = [0.0,0.550,0.699,0.799,1]

In [47]:
# Labels for the Human developmentgroups
bin_names = ['Low', 'Medium', 'High', 'Very High']

In [48]:
# Creates acidity_levels column
df_combined['human_development'] = pd.cut(df_combined['human_development'], bin_edges, labels=bin_names)

# Checks for successful creation of this column
df_combined.head()

country  year     co2  forest_coverage human_development  polity2
720  Afghanistan  1995  0.0727             2.07               Low      0.0
721      Albania  1995  0.6720            28.40            Medium      5.0
722      Algeria  1995  3.3000             0.68            Medium     -3.0
723       Angola  1995  0.7690            48.40              High     -2.0
724    Argentina  1995  3.6600            12.20              High      7.0

I will create a copy of it with only the data belonging to the last year of the data range, this one will be used to analyze the current situation. 

In [65]:
# get all columns except year
columns = [df_combined.columns.tolist()[0]] +df_combined.columns.tolist()[2:5]
columns

['country', 'co2', 'forest_coverage', 'human_development']

In [69]:
# Keep only data from 2014
df_current = df_combined.query('year == 2014')

#filter columns to drop year
df_current = df_current[columns]

#check the dataframe
df_current.head()

country    co2  forest_coverage human_development
3565  Afghanistan  0.299             2.07               Low
3566      Albania  1.960            28.20              High
3567      Algeria  3.720             0.82              High
3568       Angola  1.290            46.50               Low
3569    Argentina  4.750            10.00         Very High

I will reset the index of both dataframes because after all the data cleaning they are not in sequence anymore.

In [70]:
df_combined.reset_index(inplace =True)
df_current.reset_index(inplace =True)

The dataframes are ready to be analyzed. 

<a id='eda'></a>
## Exploratory Data Analysis

**Prepare data for analysis**
The analysis has two research questions:
1. Visualize the current situation and the relations between the indicators: for this purpose I will use only the most recent data, in the case of this dataset will be the year 2014

2. Visualize the evolution of the Co2 and the forest coverage indicators trough time: for this purpose I'll keep the dataset as formatted

### Research Question 1 (Replace this header name!)

In [50]:
df.describe()

NameError: name 'df' is not defined

In [ ]:
df['co2'].plot(kind='hist');

In [ ]:
df['forest_coverage'].plot(kind='hist');

In [ ]:
df['polity2'].plot(kind='hist');

In [ ]:
df.plot(x='co2',y='polity2',kind='scatter');

In [ ]:
 df.plot(x='forest_coverage', y='co2' ,kind='scatter')

In [ ]:
#for country in df['human_development']:
plt.scatter(y = df['human_development'], x = df['co2']);
    

In [ ]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


In [ ]:
df.groupby('human_development').mean().plot(kind='bar');


### Research Question 2  (Replace this header name!)

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work here, check over your report to make sure that it is satisfies all the areas of the rubric (found on the project submission page at the end of the lesson). You should also probably remove all of the "Tips" like this one so that the presentation is as polished as possible.

## Submitting your Project 

> Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).

> Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.

> Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [ ]:
#from subprocess import call
#call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])